File: backend/app/text_selector.py
import logging
import time

from Cocoa import NSWorkspace
from AppKit import NSPasteboard
from ApplicationServices import (
    AXUIElementCreateApplication,
    AXUIElementCopyAttributeValue,
    kAXSelectedTextAttribute,
    kAXFocusedUIElementAttribute,
)

import Quartz

logger = logging.getLogger(__name__)


class TextSelector:
    """文本选择器：Accessibility API + 剪贴板回退。"""

    def __init__(self, copy_interval: float = 0.4):
        self.last_copy_time: float = 0.0
        self.copy_interval: float = copy_interval

    def get_selected_text(self, allow_clipboard_fallback: bool = False,
                          previous_text: str = "") -> str | None:
        front_app = NSWorkspace.sharedWorkspace().frontmostApplication()
        if not front_app:
            return None
        pid = front_app.processIdentifier()
        try:
            app_ref = AXUIElementCreateApplication(pid)
            focused, err = AXUIElementCopyAttributeValue(app_ref, kAXFocusedUIElementAttribute, None)
            if err == 0 and focused:
                selected, err = AXUIElementCopyAttributeValue(focused, kAXSelectedTextAttribute, None)
                if err == 0 and selected and selected.strip():
                    return selected.strip()
        except Exception:
            logger.debug("Accessibility API 获取选中文本失败，回退到剪贴板", exc_info=True)

        if not allow_clipboard_fallback:
            return None
        return self._get_by_clipboard(previous_text)

    def _get_by_clipboard(self, previous_text: str) -> str | None:
        now = time.time()
        if now - self.last_copy_time < self.copy_interval:
            return None
        self.last_copy_time = now

        pb = NSPasteboard.generalPasteboard()
        old_text = pb.stringForType_("public.utf8-plain-text")
        old_count = pb.changeCount()
        self._simulate_cmd_c()

        new_text = None
        for _ in range(20):
            time.sleep(0.01)
            if pb.changeCount() != old_count:
                new_text = pb.stringForType_("public.utf8-plain-text")
                break

        if old_text is not None:
            pb.clearContents()
            pb.declareTypes_owner_(["public.utf8-plain-text", "org.nspasteboard.TransientType"], None)
            pb.setString_forType_(old_text, "public.utf8-plain-text")

        if new_text:
            new_text = new_text.strip()
            return new_text if new_text and new_text != previous_text else None
        return None

    @staticmethod
    def _simulate_cmd_c() -> None:
        event_down = Quartz.CGEventCreateKeyboardEvent(None, 8, True)
        Quartz.CGEventSetFlags(event_down, Quartz.kCGEventFlagMaskCommand)
        event_up = Quartz.CGEventCreateKeyboardEvent(None, 8, False)
        Quartz.CGEventPost(Quartz.kCGHIDEventTap, event_down)
        Quartz.CGEventPost(Quartz.kCGHIDEventTap, event_up)

File: Sources/AutoTranslator/TextSelector.swift
import Cocoa
import ApplicationServices
import CoreGraphics

final class TextSelector {
    private struct PasteboardItemSnapshot {
        let dataByType: [NSPasteboard.PasteboardType: Data]
    }

    private struct PasteboardSnapshot {
        let items: [PasteboardItemSnapshot]
        let fallbackString: String?
        let hadContents: Bool
    }

    private let copyInterval: TimeInterval
    private let copyPollIntervalNs: UInt64
    private let maxCopyPollCount: Int
    private var lastCopyTime: TimeInterval = 0

    init(copyInterval: TimeInterval = 0.4,
         copyPollInterval: TimeInterval = 0.01,
         maxCopyPollCount: Int = 20) {
        self.copyInterval = copyInterval
        self.copyPollIntervalNs = UInt64(copyPollInterval * 1_000_000_000)
        self.maxCopyPollCount = maxCopyPollCount
    }

    @MainActor
    func getSelectedText(allowClipboardFallback: Bool = false,
                         previousText: String = "") async -> String? {
        guard let frontApp = NSWorkspace.shared.frontmostApplication else { return nil }
        let pid = frontApp.processIdentifier

        // 尝试 Accessibility API
        if let text = getSelectedTextViaAccessibility(pid: pid) {
            return text
        }

        guard allowClipboardFallback else { return nil }
        return await getByClipboard(previousText: previousText)
    }

    private func getSelectedTextViaAccessibility(pid: pid_t) -> String? {
        let appRef = AXUIElementCreateApplication(pid)
        var focused: CFTypeRef?
        let err = AXUIElementCopyAttributeValue(appRef, kAXFocusedUIElementAttribute as CFString, &focused)
        guard err == .success, let focused = focused else { return nil }

        var selected: CFTypeRef?
        let selErr = AXUIElementCopyAttributeValue(
            focused as! AXUIElement, kAXSelectedTextAttribute as CFString, &selected
        )
        guard selErr == .success, let text = selected as? String else { return nil }
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }

    @MainActor
    private func getByClipboard(previousText: String) async -> String? {
        let now = ProcessInfo.processInfo.systemUptime
        if now - lastCopyTime < copyInterval { return nil }
        lastCopyTime = now

        let pb = NSPasteboard.general
        let snapshot = capturePasteboardSnapshot(pb)
        let oldCount = pb.changeCount
        defer { restorePasteboardSnapshot(snapshot, to: pb) }

        simulateCmdC()

        var newText: String? = nil
        for _ in 0..<maxCopyPollCount {
            if Task.isCancelled { return nil }
            try? await Task.sleep(nanoseconds: copyPollIntervalNs)
            if pb.changeCount != oldCount {
                newText = pb.string(forType: .string)
                break
            }
        }

        guard let text = newText?.trimmingCharacters(in: .whitespacesAndNewlines),
              !text.isEmpty, text != previousText else { return nil }
        return text
    }

    @MainActor
    private func capturePasteboardSnapshot(_ pasteboard: NSPasteboard) -> PasteboardSnapshot {
        let items = pasteboard.pasteboardItems ?? []
        let snapshots = items.map { item in
            var dataByType: [NSPasteboard.PasteboardType: Data] = [:]
            for type in item.types {
                if let data = item.data(forType: type) {
                    dataByType[type] = data
                }
            }
            return PasteboardItemSnapshot(dataByType: dataByType)
        }
        return PasteboardSnapshot(
            items: snapshots,
            fallbackString: pasteboard.string(forType: .string),
            hadContents: !items.isEmpty
        )
    }

    @MainActor
    private func restorePasteboardSnapshot(_ snapshot: PasteboardSnapshot, to pb: NSPasteboard) {
        pb.clearContents()
        guard snapshot.hadContents else { return }

        let restoredItems = snapshot.items.compactMap { snapshot -> NSPasteboardItem? in
            guard !snapshot.dataByType.isEmpty else { return nil }
            let item = NSPasteboardItem()
            for (type, data) in snapshot.dataByType {
                item.setData(data, forType: type)
            }
            return item
        }

        if !restoredItems.isEmpty {
            pb.writeObjects(restoredItems)
            return
        }

        guard let fallbackString = snapshot.fallbackString else { return }
        pb.declareTypes([.string, NSPasteboard.PasteboardType("org.nspasteboard.TransientType")], owner: nil)
        pb.setString(fallbackString, forType: .string)
    }

    private func simulateCmdC() {
        let src = CGEventSource(stateID: .hidSystemState)

        let down = CGEvent(keyboardEventSource: src, virtualKey: 8, keyDown: true)
        down?.flags = .maskCommand
        down?.post(tap: .cghidEventTap)

        let up = CGEvent(keyboardEventSource: src, virtualKey: 8, keyDown: false)
        up?.post(tap: .cghidEventTap)
    }
}
